batch/batch/worker/worker.py

```diff
@@ -315,91 +379,131 @@ def __init__(self, job, name, spec):
+                self.state = 'failed'
+        except asyncio.CancelledError:
+            raise
+        except Exception as e:
+            if not isinstance(e, (JobDeletedError, JobTimeoutError)):
+                log.exception(f'while running {self}')
+
+            self.state = 'error'
+            self.error = traceback.format_exc()
+        finally:
+            with self.step('deleting'):
+                await self.delete_container()
+
+    async def run_until_done_or_deleted(self, f: Callable[[], Awaitable[Any]]):
+        step = asyncio.ensure_future(f())
+        deleted = asyncio.ensure_future(self.deleted_event.wait())
+        await asyncio.wait([deleted, step], return_when=asyncio.FIRST_COMPLETED)
+        if deleted.done():
+            raise JobDeletedError()
+        assert step.done()
+        return step.result()
 
     def is_job_deleted(self) -> bool:
         return self.job.deleted
 
     def step(self, name: str):
         return self.timings.step(name)
 
-    async def get_container_status(self):
-        if not self.container:
-            return None
+    async def pull_image(self):
+        is_google_image = is_google_registry_domain(self.image_ref.domain)
+        is_public_image = self.image_ref.name() in PUBLIC_IMAGES
 
         try:
-            c = await docker_call_retry(MAX_DOCKER_OTHER_OPERATION_SECS, f'{self}')(self.container.show)
+            if not is_google_image:
+                await self.ensure_image_is_pulled()
+            elif is_public_image:
+                auth = await self.batch_worker_access_token()
+                await self.ensure_image_is_pulled(auth=auth)
+            else:
+                # Pull to verify this user has access to this
+                # image.
+                # FIXME improve the performance of this with a
+                # per-user image cache.
+                auth = self.current_user_access_token()
+                await docker_call_retry(MAX_DOCKER_IMAGE_PULL_SECS, f'{self}')(
+                    docker.images.pull, self.image_ref_str, auth=auth
+                )
         except DockerError as e:
-            if e.status == 404:
-                return None
+            if e.status == 404 and 'pull access denied' in e.message:
+                self.short_error = 'image cannot be pulled'
+            elif 'not found: manifest unknown' in e.message:
+                self.short_error = 'image not found'
             raise
 
-        cstate = c['State']
-        status = {
-            'state': cstate['Status'],
-            'started_at': cstate['StartedAt'],
-            'finished_at': cstate['FinishedAt'],
-            'out_of_memory': cstate['OOMKilled'],
-        }
-        cerror = cstate['Error']
-        if cerror:
-            status['error'] = cerror
-        else:
-            status['exit_code'] = cstate['ExitCode']
-
-        return status
+        image_config, _ = await check_exec_output('docker', 'inspect', shq(self.image_ref_str))
```

Comment:
With exec, we actually *can't* `shq`. This converts `foo;bar` into `foo\;bar` or `'foo;bar'`
```
In [2]: shq('foo;bar')                                                                                                         
Out[2]: "'foo;bar'"
```
`exec` completely avoids the shell, it passes those literal, unadulterated bytes to the program. So, we just want:
```
await check_exec_output('docker', 'inspect', self.image_ref_str)
```